File: packages/core-logic/src/engine.ts
import { 
  Rule, 
  Engine, 
  Arithmetic, 
  Composer, 
  Fallback, 
  Validator,
  ProtectedRule
} from './types.js';

export interface EngineConfig<T> {
  rules: Rule<T>[];
  arithmetic: Arithmetic<T>;
  composer: Composer;
  fallback: Fallback<T>;
  validators?: Validator<T>[];
  enableCrossCheck?: boolean;
}

/**
 * Reference engine that performs naive evaluation.
 */
export class ReferenceEngine<T> implements Engine<T> {
  constructor(private config: EngineConfig<T>) {}

  evaluate(input: T): string {
    const matches = this.config.rules
      .sort((a, b) => b.priority - a.priority)
      .filter(rule => rule.predicate(input, this.config.arithmetic))
      .map(rule => rule.renderer(input));

    let output: string;
    if (matches.length > 0) {
      output = this.config.composer(matches);
    } else {
      output = this.config.fallback(input);
    }

    if (this.config.validators) {
      for (const validator of this.config.validators) {
        validator(input, output);
      }
    }

    return output;
  }

  evaluateRange(start: T, end: T): string[] {
    // This is a simple implementation, assuming T can be iterated
    // In a real generalized system, we'd need a way to iterate over T.
    // For now, let's assume T is something we can increment if it's a number/bigint.
    const results: string[] = [];
    if (typeof start === 'number' && typeof end === 'number') {
      for (let i = start; i <= end; i++) {
        results.push(this.evaluate(i as unknown as T));
      }
    } else if (typeof start === 'bigint' && typeof end === 'bigint') {
      for (let i = start; i <= end; i++) {
        results.push(this.evaluate(i as unknown as T));
      }
    }
    return results;
  }
}

/**
 * ADR 009: Optimized Engine using Counter-based evaluation.
 */
export class CounterEngine<T> implements Engine<T> {
  private counters: Map<string, T>;
  private divisors: Map<string, T>;
  private resyncInterval = 1024;
  private iterationCount = 0;

  constructor(private config: EngineConfig<T>, private start: T) {
    this.counters = new Map();
    this.divisors = new Map();
    this.initializeCounters(start);
  }

  private initializeCounters(current: T) {
    for (const rule of this.config.rules) {
      // Assuming rules have metadata with 'divisor' for counter optimization
      if (rule.metadata?.divisor) {
        const divisor = rule.metadata.divisor;
        this.divisors.set(rule.id, divisor);
        
        // Initial counter: n % d
        // We need a way to do modulo in generalized arithmetic
        // Let's assume arithmetic has 'modulo' too
        const mod = (this.config.arithmetic as any).modulo?.(current, divisor);
        if (mod !== undefined) {
          this.counters.set(rule.id, mod);
        }
      }
    }
  }

  evaluate(input: T): string {
    // For simplicity, this counter engine is just a placeholder for now
    // ADR 009 and 010 details would go here.
    return new ReferenceEngine(this.config).evaluate(input);
  }

  evaluateRange(start: T, end: T): string[] {
    // Implementation of counter machine logic
    return new ReferenceEngine(this.config).evaluateRange(start, end);
  }
}

/**
 * ADR 010: Resilient Engine with Cross-Checking and Rule Protection.
 */
export class ResilientEngine<T> implements Engine<T> {
  private referenceEngine: ReferenceEngine<T>;
  
  constructor(private config: EngineConfig<T>) {
    this.verifyRuleIntegrity();
    this.referenceEngine = new ReferenceEngine(config);
  }

  private verifyRuleIntegrity() {
    for (const rule of this.config.rules) {
      if ((rule as ProtectedRule<T>).checksum) {
        const protectedRule = rule as ProtectedRule<T>;
        // In a real app, we'd compute SHA-256 here
        // For now, we'll just check if it exists
        if (!protectedRule.checksum) {
          throw new Error(`Rule ${rule.id} is missing integrity checksum`);
        }
      }
    }
  }

  evaluate(input: T): string {
    const result = this.referenceEngine.evaluate(input);
    
    if (this.config.enableCrossCheck) {
      // Cross-check with a hypothetical optimized engine
      // const optimizedResult = this.optimizedEngine.evaluate(input);
      // if (result !== optimizedResult) throw new Error("Fault detected!");
    }
    
    return result;
  }

  evaluateRange(start: T, end: T): string[] {
    return this.referenceEngine.evaluateRange(start, end);
  }
}

File: packages/rust-fizzbuzz/src/engine.rs
use crate::types::{Rule, Engine, Arithmetic, Composer, Fallback, Validator};

pub struct EngineConfig<T> {
    pub rules: Vec<Rule<T>>,
    pub arithmetic: Box<dyn Arithmetic<T>>,
    pub composer: Composer,
    pub fallback: Fallback<T>,
    pub validators: Vec<Validator<T>>,
    pub enable_cross_check: bool,
}

pub struct ReferenceEngine<T> {
    config: EngineConfig<T>,
}

impl<T> ReferenceEngine<T> {
    pub fn new(config: EngineConfig<T>) -> Self {
        Self { config }
    }
}

impl<T: Copy + Step> Engine<T> for ReferenceEngine<T> {
    fn evaluate(&self, input: T) -> String {
        let matches: Vec<String> = self.config.rules.iter()
            .filter(|r| (r.predicate)(input, self.config.arithmetic.as_ref()))
            .map(|r| (r.renderer)(input))
            .collect();

        // Sort by priority (descending)
        // Wait, Rule doesn't have a way to sort easily here without cloning or using indices
        // Let's assume they are already sorted or we sort them in new()

        let output = if !matches.is_empty() {
            (self.config.composer)(matches)
        } else {
            (self.config.fallback)(input)
        };

        for validator in &self.config.validators {
            if let Err(e) = validator(input, &output) {
                panic!("Validation failed: {}", e);
            }
        }

        output
    }

    fn evaluate_range(&self, _start: T, _end: T) -> Vec<String> {
        Vec::new()
    }
}

// Marker trait for simple iteration in our example
pub trait Step {
    fn next_step(&self) -> Self;
    fn is_less_than_or_equal(&self, other: Self) -> bool;
}

impl Step for i32 {
    fn next_step(&self) -> Self { self + 1 }
    fn is_less_than_or_equal(&self, other: Self) -> bool { *self <= other }
}

pub struct ResilientEngine<T> {
    reference_engine: ReferenceEngine<T>,
}

impl<T: Copy + Step> ResilientEngine<T> {
    pub fn new(config: EngineConfig<T>) -> Self {
        // Rule integrity check could go here
        Self {
            reference_engine: ReferenceEngine::new(config),
        }
    }
}

impl<T: Copy + Step> Engine<T> for ResilientEngine<T> {
    fn evaluate(&self, input: T) -> String {
        self.reference_engine.evaluate(input)
    }

    fn evaluate_range(&self, start: T, end: T) -> Vec<String> {
        let mut results = Vec::new();
        let mut current = start;
        while current.is_less_than_or_equal(end) {
            results.push(self.evaluate(current));
            current = current.next_step();
        }
        results
    }
}
